openfeature/event.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import TYPE_CHECKING, Callable, Dict, List, Optional, Union
+
+from openfeature.exception import ErrorCode
+from openfeature.provider import FeatureProvider, ProviderStatus
+
+if TYPE_CHECKING:
+    from openfeature.client import OpenFeatureClient
+
+
+class ProviderEvent(Enum):
+    PROVIDER_READY = "PROVIDER_READY"
+    PROVIDER_CONFIGURATION_CHANGED = "PROVIDER_CONFIGURATION_CHANGED"
+    PROVIDER_ERROR = "PROVIDER_ERROR"
+    PROVIDER_FATAL = "PROVIDER_FATAL"
+    PROVIDER_STALE = "PROVIDER_STALE"
+
+
+_provider_status_to_event = {
+    ProviderStatus.READY: ProviderEvent.PROVIDER_READY,
+    ProviderStatus.ERROR: ProviderEvent.PROVIDER_ERROR,
+    ProviderStatus.FATAL: ProviderEvent.PROVIDER_FATAL,
+    ProviderStatus.STALE: ProviderEvent.PROVIDER_STALE,
+}
+
+
+@dataclass
+class ProviderEventDetails:
+    flags_changed: Optional[List[str]] = None
+    message: Optional[str] = None
+    error_code: Optional[ErrorCode] = None
+    metadata: Dict[str, Union[bool, str, int, float]] = field(default_factory=dict)
+
+
+@dataclass
+class EventDetails(ProviderEventDetails):
+    provider_name: str = ""
+    flags_changed: Optional[List[str]] = None
+    message: Optional[str] = None
+    error_code: Optional[ErrorCode] = None
+    metadata: Dict[str, Union[bool, str, int, float]] = field(default_factory=dict)
+
+    @classmethod
+    def from_provider_event_details(
+        cls, provider_name: str, details: ProviderEventDetails
+    ) -> EventDetails:
+        return cls(
+            provider_name=provider_name,
+            flags_changed=details.flags_changed,
+            message=details.message,
+            error_code=details.error_code,
+            metadata=details.metadata,
+        )
+
+
+EventHandler = Callable[[EventDetails], None]
+
+
+class EventSupport:
+    _global_handlers: Dict[ProviderEvent, List[EventHandler]]
+    _client_handlers: Dict[OpenFeatureClient, Dict[ProviderEvent, List[EventHandler]]]
+
+    def __init__(self) -> None:
+        self._global_handlers = defaultdict(list)
+        self._client_handlers = defaultdict(lambda: defaultdict(list))
+
+    def run_client_handlers(
+        self, client: OpenFeatureClient, event: ProviderEvent, details: EventDetails
+    ) -> None:
+        for handler in self._client_handlers[client][event]:
```

Comment:
I don't fully understand the internals of python's hashing so maybe this isn't an issue, but it surprised me to see the client instance being used as the key here rather than some unique attribute on the client (although I'm not sure there is one!)

Reply:
Client instances hash to object identity, basically only the exact same instance in memory will match. I'll look closer into this, but apparently, it should conform to spec. Perhaps registering at the domain works just as well?

Reply:
Ah, performance should be negligible because object hashes default to using the memory address of the instance as key, which is super fast to compute.